pkg/sqs/sqs.go

```diff
@@ -18,88 +18,91 @@ limitations under the License.
-	if err != nil {
-		return nil, fmt.Errorf("error in Getting Queue URL %s", err)
-	}
+func NewAWSSqsSource(client *sqs.Client, queueURL *string) *AWSSqsSource {
 	return &AWSSqsSource{
-		queueURL:         url.QueueUrl,
-		sqsServiceClient: sqsServiceClient,
-	}, nil
+		sqsServiceClient: client,
+		queueURL:         queueURL,
+	}
 }
 
 // GetApproximateMessageCount retrieves the approximate count of messages from an AWS SQS queue based on the given queryType.
-// The queryType can be either "ApproximateNumberOfMessages" or "ApproximateNumberOfMessagesNotVisible".
+// The queryType can be either "approximateNumberOfMessages" or "ApproximateNumberOfMessagesNotVisible".
 // This function fetches the respective attribute from the SQS queue, converts it to an integer, and returns the value.
 // If an error occurs during fetching the attributes or converting the value, it logs the error and returns -1.
-func (s *AWSSqsSource) getApproximateMessageCount(queryType string) int64 {
+func (s *AWSSqsSource) getApproximateMessageCount(queryType string) (int64, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 	q := &sqs.GetQueueAttributesInput{
 		QueueUrl: s.queueURL,
-		AttributeNames: []*string{
-			aws.String(queryType),
+		AttributeNames: []types.QueueAttributeName{
+			types.QueueAttributeName(queryType),
 		},
 	}
-	attributes, err := s.sqsServiceClient.GetQueueAttributes(q)
+	attributes, err := s.sqsServiceClient.GetQueueAttributes(ctx, q)
 	if err != nil {
-		return -1
+		return -1, err
 	}
-	atoi, err := strconv.Atoi(*attributes.Attributes[queryType])
+	atoi, err := strconv.Atoi(attributes.Attributes[queryType])
 	if err != nil {
-		log.Printf("Error in Getting the Pending Items %s", err)
-		return -1
+		return -1, err
 	}
-	return int64(atoi)
+	return int64(atoi), nil
 }
 
 // Pending returns the approximate number of messages available for retrieval from the queue.
 func (s *AWSSqsSource) Pending(_ context.Context) int64 {
-	return s.getApproximateMessageCount(ApproximateNumberOfMessages)
+	count, err := s.getApproximateMessageCount(approximateNumberOfMessages)
+	if err != nil {
+		log.Fatalf("error getting approximate number of messages from queue %s", err)
+		return -1
+	}
+	return count
 }
 
 // Read fetches messages from the SQS queue and sends them to the provided message channel.
 func (s *AWSSqsSource) Read(_ context.Context, readRequest sourcesdk.ReadRequest, messageCh chan<- sourcesdk.Message) {
-	var readRequestCount int64 = 10
+	var readRequestCount int32 = 10
 
 	ctx, cancel := context.WithTimeout(context.Background(), readRequest.TimeOut())
 	defer cancel()
 	// If we have un-acked data (data which is received but yet to be deleted from the queue
-	if s.getApproximateMessageCount(ApproximateNumberOfMessagesNotVisible) > 0 {
+	count, err := s.getApproximateMessageCount(approximateNumberOfMessagesNotVisible)
```

Comment:
Does this query return the total invisible message number read by all the clients, or only by the current client? If it's the first case, it means there will be only one client working even there are multiple replicas?